### ISSUE TITLE
Crash in NCCL when running with distributed GPU 

### ISSUE THREAD

**Issue by dlwh** [google/jax#13559]
### Description

(could be a cluster issue, but unclear. This is on an HPC where I don't have sudo, and there are a lot of CUDA installations floating around, so I haven't ruled out that it's not some dumb mismatch there, but I've done my best to track that down.)

This code works fine on (1) single node with multiple gpus, or (2) pair of machines with single gpu

For multi-host, multi-gpu it hangs regardless of one gpu per node or per process.

```python
import os

import jax
import jax.numpy as jnp
import numpy as np
from jax.experimental import multihost_utils
from jax.experimental.pjit import pjit 
from jax.experimental import maps
from jax.experimental import PartitionSpec

# this is in slurm fwiw
jax.distributed.initialize(local_device_ids=[int(x) for x in os.environ.get('CUDA_VISIBLE_DEVICES', '0').split(',')])
# if slurm and one gpu per process then do
# jax.distributed.initialize()

print(jax.devices())
print(jax.process_count())

xs = jax.numpy.ones(jax.local_device_count())
r = jax.pmap(lambda x: jax.lax.psum(x, 'i'), axis_name='i')(xs)
print(r)
```

With one process per gpu and NCCL_DEBUG=TRACE I get:

```
[StreamExecutorGpuDevice(id=0, process_index=0, slice_index=0), StreamExecutorGpuDevice(id=1, process_index=1, slice_index=0), StreamExecutorGpuDevice(id=2, process_index=2, slice_index=1), StreamExecutorGpuDevice(id=3, process_index=3, slice_index=1)]
4
sphinx6:70775:70775 [0] NCCL INFO Bootstrap : Using enp49s0f0:172.24.67.98<0>
sphinx6:70775:70775 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
sphinx6:70775:70775 [0] NCCL INFO cudaDriverVersion 11070
NCCL version 2.13.4+cudaCUDA_MAJOR.CUDA_MINOR
sphinx6:70775:70775 [0] NCCL INFO NET/IB : Using [0]mlx5_0:1/IB [1]mlx5_1:1/IB [2]mlx5_2:1/IB [3]mlx5_3:1/IB [RO]; OOB enp49s0f0:172.24.67.98<0>
sphinx6:70775:70775 [0] NCCL INFO Using network IB
sphinx6:70775:70775 [0] NCCL INFO Setting affinity for GPU 0 to ff00,00000000,00000000,00000000,0000ff00
sphinx6:70775:70775 [0] NCCL INFO Channel 00/02 :    0   1   2   3
sphinx6:70775:70775 [0] NCCL INFO Channel 01/02 :    0   1   2   3
sphinx6:70775:70775 [0] NCCL INFO Trees [0] 1/2/-1->0->-1 [1] 1/-1/-1->0->2
sphinx6:70775:70775 [0] NCCL INFO Channel 00/0 : 3[87000] -> 0[a000] [receive] via NET/IB/0/GDRDMA
sphinx6:70775:70775 [0] NCCL INFO Channel 01/0 : 3[87000] -> 0[a000] [receive] via NET/IB/0/GDRDMA
sphinx6:70775:70775 [0] NCCL INFO Channel 00 : 0[a000] -> 1[47000] via P2P/IPC/read
sphinx6:70775:70775 [0] NCCL INFO Channel 01 : 0[a000] -> 1[47000] via P2P/IPC/read
sphinx6:70775:70775 [0] NCCL INFO Connected all rings
sphinx6:70775:70775 [0] NCCL INFO Channel 00/0 : 2[47000] -> 0[a000] [receive] via NET/IB/0/GDRDMA
sphinx6:70775:70775 [0] NCCL INFO Channel 01/0 : 2[47000] -> 0[a000] [receive] via NET/IB/0/GDRDMA
sphinx6:70775:70775 [0] NCCL INFO Channel 00/0 : 0[a000] -> 2[47000] [send] via NET/IB/0/GDRDMA
sphinx6:70775:70775 [0] NCCL INFO Channel 01/0 : 0[a000] -> 2[47000] [send] via NET/IB/0/GDRDMA

sphinx6:70775:71035 [0] external/nccl_archive/src/misc/ibvwrap.cc:283 NCCL WARN Call to ibv_reg_mr_iova2 failed with error Cannot allocate memory
sphinx6:70775:71035 [0] NCCL INFO external/nccl_archive/src/transport/net_ib.cc:871 -> 2
sphinx6:70775:71035 [0] NCCL INFO bazel-out/k8-opt/bin/external/nccl_archive/_virtual_includes/include_hdrs/net.h:28 -> 2
sphinx6:70775:71035 [0] NCCL INFO external/nccl_archive/src/transport/net.cc:717 -> 2
sphinx6:70775:71035 [0] NCCL INFO external/nccl_archive/src/proxy.cc:968 -> 2

sphinx6:70775:71035 [0] external/nccl_archive/src/proxy.cc:1097 NCCL WARN [Proxy Service 0] Failed to execute operation Connect from rank 0, retcode 2

sphinx6:70775:70775 [0] external/nccl_archive/src/misc/socket.cc:521 NCCL WARN Net : Connection closed by remote peer sphinx6.stanford.edu<50233>
sphinx6:70775:70775 [0] NCCL INFO external/nccl_archive/src/misc/socket.cc:529 -> 6
sphinx6:70775:70775 [0] NCCL INFO external/nccl_archive/src/misc/socket.cc:541 -> 6
sphinx6:70775:70775 [0] NCCL INFO external/nccl_archive/src/proxy.cc:859 -> 6

sphinx6:70775:70775 [0] external/nccl_archive/src/proxy.cc:862 NCCL WARN Proxy Call to rank 0 failed (Connect)
sphinx6:70775:70775 [0] NCCL INFO external/nccl_archive/src/transport/net.cc:319 -> 6
sphinx6:70775:70775 [0] NCCL INFO external/nccl_archive/src/transport.cc:134 -> 6
sphinx6:70775:70775 [0] NCCL INFO external/nccl_archive/src/init.cc:784 -> 6
sphinx6:70775:70775 [0] NCCL INFO external/nccl_archive/src/init.cc:1045 -> 6
sphinx6:70775:70775 [0] NCCL INFO external/nccl_archive/src/init.cc:1091 -> 6
sphinx6:70775:70775 [0] NCCL INFO external/nccl_archive/src/init.cc:1106 -> 6
2022-12-07 22:00:53.711817: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2153] Execution of replica 0 failed: INTERNAL: external/org_tensorflow/tensorflow/compiler/xla/service/gpu/nccl_utils.cc:285: NCCL operation ncclCommInitRank(comm.get(), nranks, id, rank) failed: remote process exited or there was a network error
Traceback (most recent call last):
  File "/juice2/u/dlwh/src/levanter/simple_distributed.py", line 19, in <module>
    r = jax.pmap(lambda x: jax.lax.psum(x, 'i'), axis_name='i')(xs)
  File "/nlp/scr/dlwh/miniconda3/envs/levanter/lib/python3.10/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/nlp/scr/dlwh/miniconda3/envs/levanter/lib/python3.10/site-packages/jax/_src/api.py", line 2256, in cache_miss
    out = map_bind_continuation(execute(*tracers))
  File "/nlp/scr/dlwh/miniconda3/envs/levanter/lib/python3.10/site-packages/jax/_src/profiler.py", line 314, in wrapper
    return func(*args, **kwargs)
  File "/nlp/scr/dlwh/miniconda3/envs/levanter/lib/python3.10/site-packages/jax/interpreters/pxla.py", line 2071, in __call__
    out_bufs = self.xla_executable.execute_sharded_on_local_devices(
jax._src.traceback_util.UnfilteredStackTrace: jaxlib.xla_extension.XlaRuntimeError: INTERNAL: external/org_tensorflow/tensorflow/compiler/xla/service/gpu/nccl_utils.cc:285: NCCL operation ncclCommInitRank(comm.get(), nranks, id, rank) failed: remote process exited or there was a network error

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/juice2/u/dlwh/src/levanter/simple_distributed.py", line 19, in <module>
    r = jax.pmap(lambda x: jax.lax.psum(x, 'i'), axis_name='i')(xs)
jaxlib.xla_extension.XlaRuntimeError: INTERNAL: external/org_tensorflow/tensorflow/compiler/xla/service/gpu/nccl_utils.cc:285: NCCL operation ncclCommInitRank(comm.get(), nranks, id, rank) failed: remote process exited or there was a network error
```

### What jax/jaxlib version are you using?

0.3.25

### Which accelerator(s) are you using?

GPU (A100s)

### Additional system info

Python 3.10.8, Ubuntu, Cuda 11.7

### NVIDIA GPU info

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 515.43.04    Driver Version: 515.43.04    CUDA Version: 11.7     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-SXM...  On   | 00000000:0A:00.0 Off |                    0 |
| N/A   28C    P0    59W / 350W |      0MiB / 81920MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM...  On   | 00000000:47:00.0 Off |                    0 |
| N/A   28C    P0    59W / 350W |      0MiB / 81920MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

which nvcc gives `/usr/local/cuda-11.7/bin/nvcc`

**Comment by mattjj** [comment]
@zhangqiaorjc or @hawkinsp does anything jump out at you as a first step here?

**Comment by dlwh** [comment]
FWIW, I now have another setup that gives a hang rather than a crash. Same basic log


```
sphinx6:95617:95730 [1] NCCL INFO Bootstrap : Using enp49s0f0:172.24.67.98<0>
sphinx6:95617:95730 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
sphinx6:95617:95728 [0] NCCL INFO cudaDriverVersion 11070
NCCL version 2.13.4+cudaCUDA_MAJOR.CUDA_MINOR
sphinx6:95617:95730 [1] NCCL INFO NET/IB : Using [0]mlx5_0:1/IB [1]mlx5_1:1/IB [2]mlx5_2:1/IB [3]mlx5_3:1/IB [RO]; OOB enp49s0f0:172.24.67.98<0>
sphinx6:95617:95730 [1] NCCL INFO Using network IB
sphinx6:95617:95728 [0] NCCL INFO Using network IB
sphinx6:95617:95730 [1] NCCL INFO Setting affinity for GPU 1 to ff00,00000000,00000000,00000000,0000ff00
sphinx6:95617:95728 [0] NCCL INFO Setting affinity for GPU 0 to ff00,00000000,00000000,00000000,0000ff00
sphinx6:95617:95728 [0] NCCL INFO Channel 00/02 :    0   1   2   3
sphinx6:95617:95728 [0] NCCL INFO Channel 01/02 :    0   1   2   3
sphinx6:95617:95730 [1] NCCL INFO Trees [0] -1/-1/-1->1->0 [1] -1/-1/-1->1->0
sphinx6:95617:95728 [0] NCCL INFO Trees [0] 1/2/-1->0->-1 [1] 1/-1/-1->0->2
sphinx6:95617:95730 [1] NCCL INFO Channel 00/0 : 1[47000] -> 2[47000] [send] via NET/IB/0(0)/GDRDMA
sphinx6:95617:95728 [0] NCCL INFO Channel 00/0 : 3[87000] -> 0[a000] [receive] via NET/IB/0/GDRDMA
sphinx6:95617:95730 [1] NCCL INFO Channel 01/0 : 1[47000] -> 2[47000] [send] via NET/IB/0(0)/GDRDMA
sphinx6:95617:95728 [0] NCCL INFO Channel 01/0 : 3[87000] -> 0[a000] [receive] via NET/IB/0/GDRDMA
sphinx6:95617:95728 [0] NCCL INFO Channel 00 : 0[a000] -> 1[47000] via P2P/direct pointer/read
sphinx6:95617:95728 [0] NCCL INFO Channel 01 : 0[a000] -> 1[47000] via P2P/direct pointer/read
sphinx6:95617:95730 [1] NCCL INFO Connected all rings
sphinx6:95617:95730 [1] NCCL INFO Channel 00 : 1[47000] -> 0[a000] via P2P/direct pointer/read
sphinx6:95617:95730 [1] NCCL INFO Channel 01 : 1[47000] -> 0[a000] via P2P/direct pointer/read
sphinx6:95617:95728 [0] NCCL INFO Connected all rings
sphinx6:95617:95728 [0] NCCL INFO Channel 00/0 : 2[47000] -> 0[a000] [receive] via NET/IB/0/GDRDMA
sphinx6:95617:95728 [0] NCCL INFO Channel 01/0 : 2[47000] -> 0[a000] [receive] via NET/IB/0/GDRDMA
sphinx6:95617:95728 [0] NCCL INFO Channel 00/0 : 0[a000] -> 2[47000] [send] via NET/IB/0/GDRDMA
sphinx6:95617:95728 [0] NCCL INFO Channel 01/0 : 0[a000] -> 2[47000] [send] via NET/IB/0/GDRDMA

sphinx6:95617:95771 [0] external/nccl_archive/src/misc/ibvwrap.cc:283 NCCL WARN Call to ibv_reg_mr_iova2 failed with error Cannot allocate memory
sphinx6:95617:95771 [0] NCCL INFO external/nccl_archive/src/transport/net_ib.cc:871 -> 2
sphinx6:95617:95771 [0] NCCL INFO bazel-out/k8-opt/bin/external/nccl_archive/_virtual_includes/include_hdrs/net.h:28 -> 2
sphinx6:95617:95771 [0] NCCL INFO external/nccl_archive/src/transport/net.cc:717 -> 2
sphinx6:95617:95771 [0] NCCL INFO external/nccl_archive/src/proxy.cc:968 -> 2
sphinx6:95617:95771 [0] NCCL INFO external/nccl_archive/src/proxy.cc:996 -> 2

sphinx6:95617:95771 [0] external/nccl_archive/src/proxy.cc:1097 NCCL WARN [Proxy Service 0] Failed to execute operation Connect from rank 0, retcode 2
sphinx6:95617:95771 [0] NCCL INFO external/nccl_archive/src/transport/net_ib.cc:690 -> 2
sphinx6:95617:95771 [0] NCCL INFO bazel-out/k8-opt/bin/external/nccl_archive/_virtual_includes/include_hdrs/net.h:27 -> 2
sphinx6:95617:95771 [0] NCCL INFO external/nccl_archive/src/transport/net.cc:643 -> 2
sphinx6:95617:95771 [0] NCCL INFO external/nccl_archive/src/proxy.cc:968 -> 2

sphinx6:95617:95771 [0] external/nccl_archive/src/proxy.cc:1097 NCCL WARN [Proxy Service 0] Failed to execute operation Connect from rank 0, retcode 2

sphinx6:95617:95728 [0] external/nccl_archive/src/misc/socket.cc:521 NCCL WARN Net : Connection closed by remote peer sphinx6.stanford.edu<34805>
sphinx6:95617:95728 [0] NCCL INFO external/nccl_archive/src/misc/socket.cc:529 -> 6
sphinx6:95617:95728 [0] NCCL INFO external/nccl_archive/src/misc/socket.cc:541 -> 6
sphinx6:95617:95728 [0] NCCL INFO external/nccl_archive/src/proxy.cc:859 -> 6

sphinx6:95617:95728 [0] external/nccl_archive/src/proxy.cc:862 NCCL WARN Proxy Call to rank 0 failed (Connect)
sphinx6:95617:95728 [0] NCCL INFO external/nccl_archive/src/transport/net.cc:319 -> 6
sphinx6:95617:95728 [0] NCCL INFO external/nccl_archive/src/transport.cc:134 -> 6
sphinx6:95617:95728 [0] NCCL INFO external/nccl_archive/src/init.cc:784 -> 6
sphinx6:95617:95728 [0] NCCL INFO external/nccl_archive/src/init.cc:1045 -> 6
sphinx6:95617:95728 [0] NCCL INFO external/nccl_archive/src/init.cc:1091 -> 6
sphinx6:95617:95728 [0] NCCL INFO external/nccl_archive/src/init.cc:1106 -> 6
```

based on lsof, it looks like libnccl is not loaded into the process? I think I saw references to vendoring libnccl, so is that what's going on? I do see that `libcudart.so.11.7.60` and `libcuda.so.515.43.04` are loaded

**Comment by hawkinsp** [comment]
I think this is a NCCL question: there's not much we're going to be able to do about this from JAX.

`nccl` is vendored by JAX at the moment, so we're using version bundled with jaxlib rather than whatever you happen to have installed.

One thing you could try is using an unvendored version of NCCL. Set the environment variable `TF_NCCL_VERSION=version` (where "version" is your version of NCCL) and build jaxlib from source. Does this help?

Do you have NCCL working successfully with any other software? If so, what version of NCCL is it using?

**Comment by hawkinsp** [comment]
It's a version number like `2.13.4`.

**Comment by dlwh** [comment]
ok yes looks like a cuda version issue. In Torch, NCCL works with 11.3 but it's not working for 11.4+. Trying to track it down on the torch side. I'll come back and update this issue with a solution for future google-ability.

**Comment by dlwh** [comment]
Ok it took some fiddling, but I got an environment set up where pytorch/deepspeed works just fine. However, I still get hangs with the jax script.

I tried disabling infiniband, but it still hangs. The outputs a bit different though:

```
(mistral4) dlwh@sc:~/src/levanter$ NCCL_IB_DISABLE=1 PATH=/usr/local/cuda-11.7/bin:$PATH XLA_FLAGS=--xla_gpu_cuda_data_dir=/usr/local/cuda-11.7 CUDA_HOME=/usr/local/cuda-11.7  NCCL_DEBUG=TRACE srun -N2 --account=nlp --partition sphinx --cpus-per-task=16 --gres=gpu:2 --job-name=dlwh-levanter-multi --mem 20g -t 10-0  --pty python simple_distributed.py
srun: job 5318854 queued and waiting for resources
srun: job 5318854 has been allocated resources
[StreamExecutorGpuDevice(id=0, process_index=0, slice_index=0), StreamExecutorGpuDevice(id=1, process_index=0, slice_index=0), StreamExecutorGpuDevice(id=2, process_index=1, slice_index=1), StreamExecutorGpuDevice(id=3, process_index=1, slice_index=1)]
2
sphinx1:2575043:2575158 [1] NCCL INFO Bootstrap : Using enp177s0f0:172.24.67.75<0>
sphinx1:2575043:2575158 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
sphinx1:2575043:2575156 [0] NCCL INFO cudaDriverVersion 11070
NCCL version 2.13.4+cudaCUDA_MAJOR.CUDA_MINOR
sphinx1:2575043:2575158 [1] NCCL INFO NCCL_IB_DISABLE set by environment to 1.
sphinx1:2575043:2575158 [1] NCCL INFO NET/Socket : Using [0]enp177s0f0:172.24.67.75<0> [1]br-0a71013f336a:192.168.192.1<0> [2]br-8138d0f1a247:192.168.128.1<0> [3]br-b66a74608043:192.168.144.1<0>
sphinx1:2575043:2575158 [1] NCCL INFO Using network Socket
sphinx1:2575043:2575156 [0] NCCL INFO Using network Socket
sphinx1:2575043:2575156 [0] NCCL INFO Setting affinity for GPU 0 to 0fc00003,00000000,00000000,00000000,0fc00003
sphinx1:2575043:2575158 [1] NCCL INFO Setting affinity for GPU 1 to 0fc00003,00000000,00000000,00000000,0fc00003
sphinx1:2575043:2575158 [1] NCCL INFO Trees [0] -1/-1/-1->1->0 [1] -1/-1/-1->1->0
sphinx1:2575043:2575156 [0] NCCL INFO Channel 00/02 :    0   1   2   3
sphinx1:2575043:2575156 [0] NCCL INFO Channel 01/02 :    0   1   2   3
sphinx1:2575043:2575156 [0] NCCL INFO Trees [0] 1/2/-1->0->-1 [1] 1/-1/-1->0->2
sphinx1:2575043:2575156 [0] NCCL INFO Channel 00/0 : 3[a000] -> 0[47000] [receive] via NET/Socket/1
sphinx1:2575043:2575156 [0] NCCL INFO Channel 01/0 : 3[a000] -> 0[47000] [receive] via NET/Socket/1
sphinx1:2575043:2575156 [0] NCCL INFO Channel 00 : 0[47000] -> 1[4d000] via P2P/direct pointer/read
sphinx1:2575043:2575158 [1] NCCL INFO Channel 00/0 : 1[4d000] -> 2[7000] [send] via NET/Socket/1
sphinx1:2575043:2575158 [1] NCCL INFO Channel 01/0 : 1[4d000] -> 2[7000] [send] via NET/Socket/1
sphinx1:2575043:2575156 [0] NCCL INFO Channel 01 : 0[47000] -> 1[4d000] via P2P/direct pointer/read
sphinx1:2575043:2575158 [1] NCCL INFO Connected all rings
sphinx1:2575043:2575158 [1] NCCL INFO Channel 00 : 1[4d000] -> 0[47000] via P2P/direct pointer/read
sphinx1:2575043:2575158 [1] NCCL INFO Channel 01 : 1[4d000] -> 0[47000] via P2P/direct pointer/read
```

after which point it hangs forever... So the warnings about not being able to allocate memory are infiniband related, but they don't seem to be the underlying issue.

**Comment by dlwh** [comment]
setting TF_NCCL_VERSION=2.13.4 didn't help

**Comment by dlwh** [comment]
I went through https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/troubleshooting.html and AFAICT the sysadmins did all the things it recommends, but I'll ask.

**Comment by dlwh** [comment]
I should mention that the slowdown from going multinode in this setup is immense (>10x). could be a programming error on my part, so it's "working" but we can't usefully use multiple nodes atm. Verified this still happens using a hand-compiled jaxlib from `main` as of today

**Comment by hawkinsp** [comment]
Does it occur with a self-built jaxlib with a newer NCCL?

**Comment by dlwh** [comment]
I'll give that a shot! It's a bit of a chore b/c AFAICT the bazel build for TF really wants nccl to be installed in the cuda directory so I have to wait on the sysadmin i think

**Comment by dlwh** [comment]
hrm, i compiled jaxlib using `TF_NCCL_VERSION=2.14.3 python build/build.py --enable_cuda --enable_nccl --cuda_path /scr-ssd/cuda-11.7/  --cudnn_version 8 --target_cpu_features release --cuda_compute_capabilities 8.0,8.6,7.5,6.1 --bazel_options "--local_ram_resources 150000"`

but when i run with NCCL_TRACE=debug it stills shows `NCCL version 2.13.4+cuda11.7`

I can confirm that bazel complains loudly if I specify the version wrong...

**Comment by dlwh** [comment]
Ok I updated to NCCL 2.16.2 (latest release) and with it and the 2.14.3 (and the manually compiled version without TF_NCCL_VERSION), I still get a failure unless I disable infiniband.

The failure is different though. It looks more like this:

```
NCCL version 2.13.4+cuda11.7
2022-12-09 20:05:25.251532: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2163] Execution of replica 0 failed: INTERNAL: Failed to execute XLA Runtime executable: run time error: custom call 'xla.gpu.all_gather' failed: Failed to get NCCL comm.
2022-12-09 20:05:35.253700: F external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2308] Replicated computation launch failed, but not all replicas terminated. Aborting process to work around deadlock. Failure message (there may have been multiple failures, see the error log for all failures):

Failed to execute XLA Runtime executable: run time error: custom call 'xla.gpu.all_gather' failed: Failed to get NCCL comm.
```

It logs much much less even with NCCL_DEBUG=TRACE on...

All versions I compiled log "NCCL version 2.13.4+cuda11.7", but I can confirm that bazel will not compile if I specify `TF_NCCL_VERSION` incorrectly, so I dunno what's going on with that version string... I nuked ~/.cache/bazel and all the bazel* files from the jax repo before rebuilding, so I don't think it's a build caching thing unless I'm missing a location.

**Comment by dlwh** [comment]
i was wrong about the less tracing. here's the log. This is with a clean conda env, clean jax repo (cleaned with git clean -fx), clean bazel cache. 

```
sphinx4:1584379:1584486 [1] NCCL INFO Bootstrap : Using enp49s0f0:172.24.67.39<0>
sphinx4:1584379:1584486 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
sphinx4:1584379:1584484 [0] NCCL INFO cudaDriverVersion 11070
NCCL version 2.13.4+cuda11.7
sphinx4:1584379:1584486 [1] NCCL INFO NET/IB : Using [0]mlx5_0:1/IB [1]mlx5_1:1/IB [2]mlx5_2:1/IB [3]mlx5_3:1/IB [RO]; OOB enp49s0f0:172.24.67.39<0>
sphinx4:1584379:1584486 [1] NCCL INFO Using network IB
sphinx4:1584379:1584484 [0] NCCL INFO Using network IB
sphinx4:1584379:1584484 [0] NCCL INFO Setting affinity for GPU 0 to 10fe,00000000,00000000,00000000,000010fe
sphinx4:1584379:1584486 [1] NCCL INFO Setting affinity for GPU 1 to 10fe,00000000,00000000,00000000,000010fe
sphinx4:1584379:1584486 [1] NCCL INFO Trees [0] -1/-1/-1->1->0 [1] -1/-1/-1->1->0
sphinx4:1584379:1584484 [0] NCCL INFO Channel 00/02 :    0   1   2   3
sphinx4:1584379:1584484 [0] NCCL INFO Channel 01/02 :    0   1   2   3
sphinx4:1584379:1584484 [0] NCCL INFO Trees [0] 1/2/-1->0->-1 [1] 1/-1/-1->0->2
sphinx4:1584379:1584484 [0] NCCL INFO Channel 00/0 : 3[c7000] -> 0[7000] [receive] via NET/IB/0/GDRDMA
sphinx4:1584379:1584484 [0] NCCL INFO Channel 01/0 : 3[c7000] -> 0[7000] [receive] via NET/IB/0/GDRDMA
sphinx4:1584379:1584484 [0] NCCL INFO Channel 00 : 0[7000] -> 1[4d000] via P2P/direct pointer/read
sphinx4:1584379:1584484 [0] NCCL INFO Channel 01 : 0[7000] -> 1[4d000] via P2P/direct pointer/read
sphinx4:1584379:1584486 [1] NCCL INFO Channel 00/0 : 1[4d000] -> 2[a000] [send] via NET/IB/0(0)/GDRDMA
sphinx4:1584379:1584486 [1] NCCL INFO Channel 01/0 : 1[4d000] -> 2[a000] [send] via NET/IB/0(0)/GDRDMA
sphinx4:1584379:1584484 [0] NCCL INFO Connected all rings
sphinx4:1584379:1584486 [1] NCCL INFO Connected all rings
sphinx4:1584379:1584486 [1] NCCL INFO Channel 00 : 1[4d000] -> 0[7000] via P2P/direct pointer/read
sphinx4:1584379:1584486 [1] NCCL INFO Channel 01 : 1[4d000] -> 0[7000] via P2P/direct pointer/read
sphinx4:1584379:1584484 [0] NCCL INFO Channel 00/0 : 2[a000] -> 0[7000] [receive] via NET/IB/0/GDRDMA
sphinx4:1584379:1584484 [0] NCCL INFO Channel 01/0 : 2[a000] -> 0[7000] [receive] via NET/IB/0/GDRDMA
sphinx4:1584379:1584484 [0] NCCL INFO Channel 00/0 : 0[7000] -> 2[a000] [send] via NET/IB/0/GDRDMA
sphinx4:1584379:1584484 [0] NCCL INFO Channel 01/0 : 0[7000] -> 2[a000] [send] via NET/IB/0/GDRDMA

sphinx4:1584379:1584541 [0] misc/ibvwrap.cc:283 NCCL WARN Call to ibv_reg_mr_iova2 failed with error Cannot allocate memory
sphinx4:1584379:1584541 [0] NCCL INFO transport/net_ib.cc:871 -> 2
sphinx4:1584379:1584541 [0] NCCL INFO include/net.h:28 -> 2
sphinx4:1584379:1584541 [0] NCCL INFO transport/net.cc:717 -> 2
sphinx4:1584379:1584541 [0] NCCL INFO proxy.cc:968 -> 2

sphinx4:1584379:1584541 [0] proxy.cc:1097 NCCL WARN [Proxy Service 0] Failed to execute operation Connect from rank 0, retcode 2

sphinx4:1584379:1584484 [0] misc/socket.cc:521 NCCL WARN Net : Connection closed by remote peer sphinx4.stanford.edu<44363>
sphinx4:1584379:1584484 [0] NCCL INFO misc/socket.cc:529 -> 6
sphinx4:1584379:1584484 [0] NCCL INFO misc/socket.cc:541 -> 6
sphinx4:1584379:1584484 [0] NCCL INFO proxy.cc:859 -> 6

sphinx4:1584379:1584484 [0] proxy.cc:862 NCCL WARN Proxy Call to rank 0 failed (Connect)
sphinx4:1584379:1584484 [0] NCCL INFO transport/net.cc:319 -> 6
sphinx4:1584379:1584484 [0] NCCL INFO transport.cc:134 -> 6
sphinx4:1584379:1584484 [0] NCCL INFO init.cc:784 -> 6
sphinx4:1584379:1584484 [0] NCCL INFO init.cc:1045 -> 6
sphinx4:1584379:1584484 [0] NCCL INFO init.cc:1091 -> 6
sphinx4:1584379:1584484 [0] NCCL INFO init.cc:1106 -> 6
2022-12-12 11:29:16.913408: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2163] Execution of replica 0 failed: INTERNAL: Failed to execute XLA Runtime executable: run time error: custom call 'xla.gpu.all_reduce_start' failed: Failed to get NcclComm.
2022-12-12 11:29:26.913593: F external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2308] Replicated computation launch failed, but not all replicas terminated. Aborting process to work around deadlock. Failure message (there may have been multiple failures, see the error log for all failures):

Failed to execute XLA Runtime executable: run time error: custom call 'xla.gpu.all_reduce_start' failed: Failed to get NcclComm.
srun: error: sphinx4: task 0: Aborted
srun: error: sphinx5: task 1: Aborted (core dumped)
```

**Comment by dlwh** [comment]
I realized that when you use TF_NCCL_VERSION it wants to dynamically load the .so . Once I specified the LD_LIBRARY_PATH it loads the right NCCL, but I'm now running into driver version issues. gonna work through that.

**Comment by dlwh** [comment]
Ok the issue is still present in 2.14.3 . There are two newer versions of nccl but they require newer cuda drivers than we have installed. I will update once we're able to update the cuda drivers to 11.8. Requires cluster downtime so it's not a trivial thing.

**Comment by dlwh** [comment]
ok i believe have separated this issue from infiniband: using two nodes that have 3090s (no IB) I get the same failure unless I restrict the connection to use only ethernet and not the 'br' interfaces.

**Comment by dlwh** [comment]
Hi @yhtang, all, just wanted to redraw attention to this since it might have gotten lost over the holidays.

I'm also happy to workaround by installing different cuda drivers and things. Please let me know if there's a "known good" set of versions of NCCL/CUDA/Jax/etc that works...

**Comment by dlwh** [comment]
I believe that it dramatically hurts throughput to do this, fwiw, so it's more of a workaround than a solution

**Comment by tchanah** [comment]
Is there any better solution?

**Comment by dlwh** [comment]
Not that I’m aware of :(

**Comment by dlwh** [comment]
this is fixed with latest jax (0.4.13), cuda (12.1.1), cudnn (8.9)

**Comment by foursunZero** [comment]
> this is fixed with latest jax (0.4.13), cuda (12.1.1), cudnn (8.9)

So Is it really upgrade version to solve the problem? latest jax (0.4.13), cuda (12.1.1), cudnn (8.9) can't help me. I haven't fix it.
